internal/tools/aws/client.go

```diff
@@ -24,43 +39,72 @@ type AWS interface {
 
 	TagResource(resourceID, key, value string, logger log.FieldLogger) error
 	UntagResource(resourceID, key, value string, logger log.FieldLogger) error
-	IsValidAMI(AMIImage string) (bool, error)
+	IsValidAMI(AMIImage string, logger log.FieldLogger) (bool, error)
+}
+
+// NewAWSClientWithConfig returns a new instance of Client with a custom configuration.
+func NewAWSClientWithConfig(config *aws.Config) *Client {
+	return &Client{
+		config: config,
+		mux:    &sync.Mutex{},
+	}
+}
+
+// Service hold AWS clients for each service.
+type Service struct {
+	acm            acmiface.ACMAPI
+	ec2            ec2iface.EC2API
+	iam            iamiface.IAMAPI
+	rds            rdsiface.RDSAPI
+	s3             s3iface.S3API
+	route53        route53iface.Route53API
+	secretsManager secretsmanageriface.SecretsManagerAPI
+}
+
+// NewService creates a new instance of Service.
+func NewService(sess *session.Session) *Service {
+	return &Service{
+		acm:            acm.New(sess),
+		iam:            iam.New(sess),
+		rds:            rds.New(sess),
+		s3:             s3.New(sess),
+		route53:        route53.New(sess),
+		secretsManager: secretsmanager.New(sess),
+		ec2:            ec2.New(sess),
+	}
 }
 
 // Client is a client for interacting with AWS resources.
 type Client struct {
-	api   api
-	store model.InstallationDatabaseStoreInterface
+	store   model.InstallationDatabaseStoreInterface
+	service *Service
+	config  *aws.Config
+	mux     *sync.Mutex
 }
 
-// api mocks out the AWS API calls for testing.
-type api interface {
-	getRoute53Client() (*route53.Route53, error)
-	changeResourceRecordSets(*route53.Route53, *route53.ChangeResourceRecordSetsInput) (*route53.ChangeResourceRecordSetsOutput, error)
-	listResourceRecordSets(*route53.Route53, *route53.ListResourceRecordSetsInput) (*route53.ListResourceRecordSetsOutput, error)
-	listHostedZones(*route53.Route53, *route53.ListHostedZonesInput) (*route53.ListHostedZonesOutput, error)
-	listTagsForResource(*route53.Route53, *route53.ListTagsForResourceInput) (*route53.ListTagsForResourceOutput, error)
-
-	getEC2Client() (*ec2.EC2, error)
-	tagResource(*ec2.EC2, *ec2.CreateTagsInput) (*ec2.CreateTagsOutput, error)
-	untagResource(*ec2.EC2, *ec2.DeleteTagsInput) (*ec2.DeleteTagsOutput, error)
-	describeImages(svc *ec2.EC2, input *ec2.DescribeImagesInput) (*ec2.DescribeImagesOutput, error)
-
-	getACMClient() (*acm.ACM, error)
-	listCertificates(*acm.ACM, *acm.ListCertificatesInput) (*acm.ListCertificatesOutput, error)
-	listTagsForCertificate(*acm.ACM, *acm.ListTagsForCertificateInput) (*acm.ListTagsForCertificateOutput, error)
-}
+// Service contructs an AWS session if not yet successfully done and returns AWS clients.
+func (c *Client) Service(logger log.FieldLogger) *Service {
```

Comment:
Is there a reason that `logger` is not part of the `Client` struct, and is instead passed into this method every time? This pattern creates a lot of repetition in other files, where I do not see the extra power gained by including it as an argument utilized, which simply makes calls to this method more verbose. I think we should either take advantage of the call sites of this method by providing `logger` arguments that add more context from those call sites or eliminate this argument and set a `logger` per `Client` that can be manually overridden if necessary to add more context since we are not doing that in the majority of call sites of this method.

Reply:
https://github.com/mattermost/mattermost-cloud/pull/159#discussion_r389662499

Reply:
See my other comment, then, the behavior here is incorrect if the logger is passed in. Only the first one supplied per `Client` is actually effective.

I'd like to see extra human-friendly context added to the logger that's passed in, if we are going to the trouble to pass the logger in each time, too, to aid in debugging, since we create a new logger for each call site. Otherwise we're repeating ourselves for no reason, but if there's unique context added at each call site, that would be a good reason to pass the logger in as an argument and will aid with debugging. 

@gabrieljackson can weigh in if I'm totally off base here, since I'm still getting used to this pattern myself.